File: src/astroviper/core/imaging/gridders/prolate_spheroidal_grid.py
from numba import jit
import numpy as np
import numpy.typing as npt
import math


# When jit is used round is repolaced by standard c++ round that is different to python round
@jit(nopython=True, cache=True, nogil=True)  # fastmath=True
def prolate_spheroidal_grid_jit(
    grid,
    normalization,
    vis_data,
    uvw,
    frequency_coord,
    frequency_map,
    time_map,
    pol_map,
    weight,
    cgk_1D,
    n_uv,
    delta_lm,
    support,
    oversampling,
):
    """Grid weighted visibilities onto a UV plane using a prolate spheroidal convolution kernel.

    Accumulates weighted, convolved visibility samples into ``grid`` and the
    corresponding sum of convolution weights into ``normalization``.  Both
    arrays are modified in place so that repeated calls (e.g. over multiple
    measurement sets) accumulate contributions.

    Time and frequency mapping allow the input visibility axes to be collapsed
    or reordered onto the image axes.

    Convention note — axis prefixes:
      * ``m_`` denotes an image-domain axis size.
      * ``n_`` denotes a visibility-domain axis size.

    Parameters
    ----------
    grid : complex ndarray, shape (m_time, m_chan, m_pol, m_u, m_v)
        UV grid accumulator.  Modified in place.
    normalization : float ndarray, shape (m_time, m_chan, m_pol)
        Per-cell sum of ``weight * convolution_kernel`` values, used to
        normalise the grid after all visibilities have been accumulated.
        Modified in place.
    vis_data : complex ndarray, shape (n_time, n_baseline, n_vis_chan, n_pol)
        Correlated visibility data.
    uvw : float ndarray, shape (n_time, n_baseline, 3)
        Baseline coordinates in metres.  The third axis is ordered (U, V, W).
    frequency_coord : float ndarray, shape (n_vis_chan,)
        Channel centre frequencies in Hz, used to convert UVW from metres to
        wavelengths.
    frequency_map : int ndarray, shape (n_vis_chan,)
        Maps each visibility channel index to its target image channel index
        ``a_chan`` in ``grid``.
    time_map : int ndarray, shape (n_time,)
        Maps each visibility time index to its target image time index
        ``a_time`` in ``grid``.
    pol_map : int ndarray, shape (n_pol,)
        Maps each visibility polarization index to its target image
        polarization index ``a_pol`` in ``grid``.
    weight : float ndarray, shape (n_time, n_baseline, n_vis_chan, n_pol)
        Imaging weights applied to each visibility sample before gridding.
        Samples with ``NaN`` weight or zero weighted data are skipped.
    cgk_1D : float ndarray, shape (oversampling * (support // 2 + 1),)
        Oversampled 1-D prolate spheroidal wave function (PSWF) kernel.
        The 2-D kernel is formed as the outer product of two 1-D lookups.
    n_uv : int ndarray, shape (2,)
        Padded UV-grid dimensions ``[n_u, n_v]`` in pixels.
    delta_lm : float ndarray, shape (2,)
        Image cell size in radians ``[delta_l, delta_m]``.
    support : int
        Full convolution support width in grid pixels.  The kernel spans
        ``[-support//2, support - support//2)`` pixels around each sample.
    oversampling : int
        Oversampling factor of the 1-D kernel ``cgk_1D``.  A higher value
        reduces interpolation error at the cost of a larger kernel table.

    Returns
    -------
    None
        ``grid`` and ``normalization`` are modified in place; nothing is
        returned.

    Notes
    -----
    - Visibilities whose U or V coordinate is ``NaN`` are skipped entirely.
    - Samples whose weighted data value is ``NaN`` or zero are skipped.
    - The sub-pixel offset from the nearest grid point is rounded using
      ``int(x + 0.5)`` rather than ``numpy.round`` to match Fortran/C++
      rounding behaviour (round-half-away-from-zero).
    - Hardcoding ``support`` and ``oversampling`` as literals (currently
      commented out) allows Numba to unroll the inner support loops,
      yielding a significant speed-up.
    """

    # By hardcoding the support and oversampling values, the innermost for loops can be unrolled by the compiler leading to significantly faster code.
    # support = 7
    # oversampling = 100

    c = 299792458.0
    uv_scale = np.zeros((2, len(frequency_coord)), dtype=np.double)
    uv_scale[0, :] = -(frequency_coord * delta_lm[0] * n_uv[0]) / c
    uv_scale[1, :] = -(frequency_coord * delta_lm[1] * n_uv[1]) / c

    # oversampling_center = int(oversampling // 2)
    support_center = int(support // 2)
    uv_center = n_uv // 2

    start_support = -support_center
    end_support = (
        support - support_center
    )  # end_support is larger by 1 so that python range() gives correct indices

    n_time = uvw.shape[0]
    n_baseline = uvw.shape[1]
    n_chan = len(frequency_map)
    n_pol = len(pol_map)

    n_u = n_uv[0]
    n_v = n_uv[1]

    for i_time in range(n_time):
        a_time = time_map[i_time]
        for i_baseline in range(n_baseline):
            for i_chan in range(n_chan):
                a_chan = frequency_map[i_chan]
                u = uvw[i_time, i_baseline, 0] * uv_scale[0, i_chan]
                v = uvw[i_time, i_baseline, 1] * uv_scale[1, i_chan]

                if ~np.isnan(u) and ~np.isnan(v):
                    u_pos = u + uv_center[0]
                    v_pos = v + uv_center[1]

                    # u_pos_conj = -u + uv_center[0]
                    # v_pos_conj = -v + uv_center[1]

                    # Doing round as int(x+0.5) since u_pos/v_pos should always positive and this matices fortran and gives consistant rounding.
                    # u_center_indx = int(u_pos + 0.5)
                    # v_center_indx = int(v_pos + 0.5)

                    # Do not use numpy round
                    u_center_indx = int(u_pos + 0.5)
                    v_center_indx = int(v_pos + 0.5)

                    if (
                        (u_center_indx + support_center < n_u)
                        and (v_center_indx + support_center < n_v)
                        and (u_center_indx - support_center >= 0)
                        and (v_center_indx - support_center >= 0)
                    ):
                        u_offset = u_center_indx - u_pos
                        u_center_offset_indx = math.floor(u_offset * oversampling + 0.5)
                        v_offset = v_center_indx - v_pos
                        v_center_offset_indx = math.floor(v_offset * oversampling + 0.5)

                        for i_pol in range(n_pol):
                            sel_weight = weight[i_time, i_baseline, i_chan, i_pol]
                            weighted_data = (
                                vis_data[i_time, i_baseline, i_chan, i_pol]
                                * weight[i_time, i_baseline, i_chan, i_pol]
                            )

                            if ~np.isnan(weighted_data) and (weighted_data != 0.0):
                                a_pol = pol_map[i_pol]
                                norm = 0.0

                                for i_v in range(start_support, end_support):
                                    v_indx = v_center_indx + i_v
                                    v_offset_indx = np.abs(
                                        oversampling * i_v + v_center_offset_indx
                                    )
                                    conv_v = cgk_1D[v_offset_indx]

                                    for i_u in range(start_support, end_support):
                                        u_indx = u_center_indx + i_u
                                        u_offset_indx = np.abs(
                                            oversampling * i_u + u_center_offset_indx
                                        )
                                        conv_u = cgk_1D[u_offset_indx]
                                        conv = conv_u * conv_v

                                        grid[a_time, a_chan, a_pol, u_indx, v_indx] = (
                                            grid[a_time, a_chan, a_pol, u_indx, v_indx]
                                            + conv * weighted_data
                                        )
                                        norm = norm + conv

                                normalization[a_time, a_chan, a_pol] = (
                                    normalization[a_time, a_chan, a_pol]
                                    + sel_weight * norm
                                )
    return


# Gridder used to create PSF (UV-sampling grid / point spread function)
# When jit is used round is repolaced by standard c++ round that is different to python round
@jit(nopython=True, cache=True, nogil=True)  # fastmath=True
def prolate_spheroidal_grid_uv_sampling_jit(
    grid,
    normalization,
    uvw,
    frequency_coord,
    frequency_map,
    time_map,
    pol_map,
    weight,
    cgk_1D,
    n_uv,
    delta_lm,
    support,
    oversampling,
):
    """Grid imaging weights onto a UV plane to form the UV-sampling function (PSF numerator).

    This is the PSF variant of :func:`prolate_spheroidal_grid_jit`.  Instead
    of gridding weighted visibility data, it grids the imaging weights alone,
    accumulating the UV-sampling function that describes the PSF of the
    interferometric array.  The result is used to form the dirty beam (point
    spread function) and to normalise the dirty image.

    Both ``grid`` and ``normalization`` are modified in place so that repeated
    calls over multiple measurement sets accumulate contributions.

    Convention note — axis prefixes:
      * ``m_`` denotes an image-domain axis size.
      * ``n_`` denotes a visibility-domain axis size.

    Parameters
    ----------
    grid : complex ndarray, shape (m_time, m_chan, m_pol, m_u, m_v)
        UV-sampling grid accumulator.  Modified in place.
    normalization : float ndarray, shape (m_time, m_chan, m_pol)
        Per-cell sum of ``weight * convolution_kernel`` values.  Modified in
        place.
    uvw : float ndarray, shape (n_time, n_baseline, 3)
        Baseline coordinates in metres.  The third axis is ordered (U, V, W).
    frequency_coord : float ndarray, shape (n_vis_chan,)
        Channel centre frequencies in Hz, used to convert UVW from metres to
        wavelengths.
    frequency_map : int ndarray, shape (n_vis_chan,)
        Maps each visibility channel index to its target image channel index
        ``a_chan`` in ``grid``.
    time_map : int ndarray, shape (n_time,)
        Maps each visibility time index to its target image time index
        ``a_time`` in ``grid``.
    pol_map : int ndarray, shape (n_pol,)
        Maps each visibility polarization index to its target image
        polarization index ``a_pol`` in ``grid``.
    weight : float ndarray, shape (n_time, n_baseline, n_vis_chan, n_pol)
        Imaging weights.  Samples with ``NaN`` weight or zero weight are
        skipped.
    cgk_1D : float ndarray, shape (oversampling * (support // 2 + 1),)
        Oversampled 1-D prolate spheroidal wave function (PSWF) kernel.
        The 2-D kernel is formed as the outer product of two 1-D lookups.
    n_uv : int ndarray, shape (2,)
        Padded UV-grid dimensions ``[n_u, n_v]`` in pixels.
    delta_lm : float ndarray, shape (2,)
        Image cell size in radians ``[delta_l, delta_m]``.
    support : int
        Full convolution support width in grid pixels.  The kernel spans
        ``[-support//2, support - support//2)`` pixels around each sample.
    oversampling : int
        Oversampling factor of the 1-D kernel ``cgk_1D``.

    Returns
    -------
    None
        ``grid`` and ``normalization`` are modified in place; nothing is
        returned.

    Notes
    -----
    - Visibilities whose U or V coordinate is ``NaN`` are skipped entirely.
    - Samples with ``NaN`` or zero weight are skipped.
    - Sub-pixel offsets are rounded with ``int(x + 0.5)`` rather than
      ``numpy.round`` to match Fortran/C++ rounding behaviour
      (round-half-away-from-zero).
    - Hardcoding ``support`` and ``oversampling`` as literals (currently
      commented out) allows Numba to unroll the inner support loops,
      yielding a significant speed-up.
    """

    # By hardcoding the support and oversampling values, the innermost for loops can be unrolled by the compiler leading to significantly faster code.
    # support = 7
    # oversampling = 100

    c = 299792458.0
    uv_scale = np.zeros((2, len(frequency_coord)), dtype=np.double)
    uv_scale[0, :] = -(frequency_coord * delta_lm[0] * n_uv[0]) / c
    uv_scale[1, :] = -(frequency_coord * delta_lm[1] * n_uv[1]) / c

    # oversampling_center = int(oversampling // 2)
    support_center = int(support // 2)
    uv_center = n_uv // 2

    start_support = -support_center
    end_support = (
        support - support_center
    )  # end_support is larger by 1 so that python range() gives correct indices

    n_time = uvw.shape[0]
    n_baseline = uvw.shape[1]
    n_chan = len(frequency_map)
    n_pol = len(pol_map)

    n_u = n_uv[0]
    n_v = n_uv[1]

    for i_time in range(n_time):
        a_time = time_map[i_time]
        for i_baseline in range(n_baseline):
            for i_chan in range(n_chan):
                a_chan = frequency_map[i_chan]
                u = uvw[i_time, i_baseline, 0] * uv_scale[0, i_chan]
                v = uvw[i_time, i_baseline, 1] * uv_scale[1, i_chan]

                if ~np.isnan(u) and ~np.isnan(v):
                    u_pos = u + uv_center[0]
                    v_pos = v + uv_center[1]

                    u_pos_conj = -u + uv_center[0]
                    v_pos_conj = -v + uv_center[1]

                    # Doing round as int(x+0.5) since u_pos/v_pos should always positive and this matices fortran and gives consistant rounding.
                    # u_center_indx = int(u_pos + 0.5)
                    # v_center_indx = int(v_pos + 0.5)

                    # Do not use numpy round
                    u_center_indx = int(u_pos + 0.5)
                    v_center_indx = int(v_pos + 0.5)

                    u_center_indx_conj = int(u_pos_conj + 0.5)
                    v_center_indx_conj = int(v_pos_conj + 0.5)

                    if (
                        (u_center_indx + support_center < n_u)
                        and (v_center_indx + support_center < n_v)
                        and (u_center_indx - support_center >= 0)
                        and (v_center_indx - support_center >= 0)
                    ):
                        u_offset = u_center_indx - u_pos
                        u_center_offset_indx = math.floor(u_offset * oversampling + 0.5)
                        v_offset = v_center_indx - v_pos
                        v_center_offset_indx = math.floor(v_offset * oversampling + 0.5)

                        for i_pol in range(n_pol):
                            weight_data = weight[i_time, i_baseline, i_chan, i_pol]

                            if ~np.isnan(weight_data) and (weight_data != 0.0):
                                a_pol = pol_map[i_pol]
                                norm = 0.0

                                for i_v in range(start_support, end_support):
                                    v_indx = v_center_indx + i_v
                                    v_offset_indx = np.abs(
                                        oversampling * i_v + v_center_offset_indx
                                    )
                                    conv_v = cgk_1D[v_offset_indx]

                                    for i_u in range(start_support, end_support):
                                        u_indx = u_center_indx + i_u
                                        u_offset_indx = np.abs(
                                            oversampling * i_u + u_center_offset_indx
                                        )
                                        conv_u = cgk_1D[u_offset_indx]
                                        conv = conv_u * conv_v

                                        grid[a_time, a_chan, a_pol, u_indx, v_indx] = (
                                            grid[a_time, a_chan, a_pol, u_indx, v_indx]
                                            + conv * weight_data
                                        )
                                        norm = norm + conv

                                normalization[a_time, a_chan, a_pol] = (
                                    normalization[a_time, a_chan, a_pol]
                                    + weight_data * norm
                                )
    return

File: src/astroviper/core/imaging/gridders/prolate_spheroidal_grid_cpp/src/prolate_spheroidal_grid.cpp
#include "../include/prolate_spheroidal_grid.hpp"

#include <cmath>
#include <complex>
#include <vector>

namespace prolate_spheroidal {

void prolate_spheroidal_grid(
    std::complex<double>* grid,
    double* normalization,
    const std::complex<double>* vis_data,
    const double* uvw,
    const double* frequency_coord,
    const int64_t* frequency_map,
    const int64_t* time_map,
    const int64_t* pol_map,
    const double* weight,
    const double* cgk_1D,
    int m_time_g, int m_chan_g, int m_pol_g, int m_u, int m_v,
    int n_time, int n_baseline, int n_vis_chan, int n_pol,
    double delta_l, double delta_m,
    int support, int oversampling
) {
    constexpr double c = 299792458.0;

    // Precompute per-channel UV scaling factors:
    //   uv_scale[0, i] = -(freq[i] * delta_l * m_u) / c
    //   uv_scale[1, i] = -(freq[i] * delta_m * m_v) / c
    std::vector<double> uv_scale_u(n_vis_chan), uv_scale_v(n_vis_chan);
    for (int i = 0; i < n_vis_chan; ++i) {
        uv_scale_u[i] = -(frequency_coord[i] * delta_l * m_u) / c;
        uv_scale_v[i] = -(frequency_coord[i] * delta_m * m_v) / c;
    }

    const int support_center = support / 2;
    const int uv_center_u    = m_u / 2;
    const int uv_center_v    = m_v / 2;
    const int start_support  = -support_center;
    const int end_support    = support - support_center;

    // Grid strides for shape (m_time_g, m_chan_g, m_pol_g, m_u, m_v):
    //   grid[a_time, a_chan, a_pol, u_idx, v_idx] =
    //     grid[((a_time*m_chan_g + a_chan)*m_pol_g + a_pol)*m_u*m_v + u_idx*m_v + v_idx]
    const int grid_pol_stride  = m_u * m_v;
    const int grid_chan_stride  = m_pol_g * grid_pol_stride;
    const int grid_time_stride  = m_chan_g * grid_chan_stride;

    // Normalization strides for shape (m_time_g, m_chan_g, m_pol_g):
    //   norm[a_time, a_chan, a_pol] = norm[(a_time*m_chan_g + a_chan)*m_pol_g + a_pol]
    const int norm_chan_stride = m_pol_g;
    const int norm_time_stride = m_chan_g * norm_chan_stride;

    // Visibility / weight strides for shape (n_time, n_baseline, n_vis_chan, n_pol):
    //   arr[i_time, i_baseline, i_chan, i_pol] =
    //     arr[((i_time*n_baseline + i_baseline)*n_vis_chan + i_chan)*n_pol + i_pol]
    const int vis_pol_stride  = 1;
    const int vis_chan_stride  = n_pol;
    const int vis_bl_stride   = n_vis_chan * vis_chan_stride;
    const int vis_time_stride = n_baseline * vis_bl_stride;

    // UVW strides for shape (n_time, n_baseline, 3):
    const int uvw_bl_stride   = 3;
    const int uvw_time_stride = n_baseline * uvw_bl_stride;

    for (int i_time = 0; i_time < n_time; ++i_time) {
        const int a_time    = time_map[i_time];
        const int vis_t_off = i_time * vis_time_stride;
        const int uvw_t_off = i_time * uvw_time_stride;

        for (int i_baseline = 0; i_baseline < n_baseline; ++i_baseline) {
            const int vis_b_off = vis_t_off + i_baseline * vis_bl_stride;
            const int uvw_b_off = uvw_t_off + i_baseline * uvw_bl_stride;

            for (int i_chan = 0; i_chan < n_vis_chan; ++i_chan) {
                const int a_chan = frequency_map[i_chan];

                const double u = uvw[uvw_b_off + 0] * uv_scale_u[i_chan];
                const double v = uvw[uvw_b_off + 1] * uv_scale_v[i_chan];

                if (std::isnan(u) || std::isnan(v)) continue;

                const double u_pos = u + uv_center_u;
                const double v_pos = v + uv_center_v;

                // Round half-away-from-zero: int(x + 0.5) matches Fortran/C++ convention
                const int u_center_indx = static_cast<int>(u_pos + 0.5);
                const int v_center_indx = static_cast<int>(v_pos + 0.5);

                if (u_center_indx + support_center >= m_u) continue;
                if (v_center_indx + support_center >= m_v) continue;
                if (u_center_indx - support_center <  0)   continue;
                if (v_center_indx - support_center <  0)   continue;

                // Sub-pixel offset and its oversampled index
                const double u_offset = u_center_indx - u_pos;
                const int    u_center_offset_indx =
                    static_cast<int>(std::floor(u_offset * oversampling + 0.5));

                const double v_offset = v_center_indx - v_pos;
                const int    v_center_offset_indx =
                    static_cast<int>(std::floor(v_offset * oversampling + 0.5));

                const int vis_c_off = vis_b_off + i_chan * vis_chan_stride;

                for (int i_pol = 0; i_pol < n_pol; ++i_pol) {
                    const int vis_idx = vis_c_off + i_pol;

                    const double sel_weight = weight[vis_idx];
                    const std::complex<double> wd = vis_data[vis_idx] * sel_weight;

                    if (std::isnan(wd.real()) || std::isnan(wd.imag())) continue;
                    if (wd == std::complex<double>(0.0, 0.0)) continue;

                    const int a_pol = pol_map[i_pol];

                    const int grid_base = a_time * grid_time_stride
                                        + a_chan  * grid_chan_stride
                                        + a_pol   * grid_pol_stride;
                    const int norm_idx  = a_time * norm_time_stride
                                        + a_chan  * norm_chan_stride
                                        + a_pol;

                    double norm = 0.0;

                    for (int i_v = start_support; i_v < end_support; ++i_v) {
                        const int v_indx = v_center_indx + i_v;
                        const int v_offset_indx =
                            std::abs(oversampling * i_v + v_center_offset_indx);
                        const double conv_v = cgk_1D[v_offset_indx];

                        for (int i_u = start_support; i_u < end_support; ++i_u) {
                            const int u_indx = u_center_indx + i_u;
                            const int u_offset_indx =
                                std::abs(oversampling * i_u + u_center_offset_indx);
                            const double conv = cgk_1D[u_offset_indx] * conv_v;

                            grid[grid_base + u_indx * m_v + v_indx] += conv * wd;
                            norm += conv;
                        }
                    }

                    normalization[norm_idx] += sel_weight * norm;
                }
            }
        }
    }
}


void prolate_spheroidal_grid_uv_sampling(
    std::complex<double>* grid,
    double* normalization,
    const double* uvw,
    const double* frequency_coord,
    const int64_t* frequency_map,
    const int64_t* time_map,
    const int64_t* pol_map,
    const double* weight,
    const double* cgk_1D,
    int m_time_g, int m_chan_g, int m_pol_g, int m_u, int m_v,
    int n_time, int n_baseline, int n_vis_chan, int n_pol,
    double delta_l, double delta_m,
    int support, int oversampling
) {
    constexpr double c = 299792458.0;

    std::vector<double> uv_scale_u(n_vis_chan), uv_scale_v(n_vis_chan);
    for (int i = 0; i < n_vis_chan; ++i) {
        uv_scale_u[i] = -(frequency_coord[i] * delta_l * m_u) / c;
        uv_scale_v[i] = -(frequency_coord[i] * delta_m * m_v) / c;
    }

    const int support_center = support / 2;
    const int uv_center_u    = m_u / 2;
    const int uv_center_v    = m_v / 2;
    const int start_support  = -support_center;
    const int end_support    = support - support_center;

    const int grid_pol_stride  = m_u * m_v;
    const int grid_chan_stride  = m_pol_g * grid_pol_stride;
    const int grid_time_stride  = m_chan_g * grid_chan_stride;

    const int norm_chan_stride = m_pol_g;
    const int norm_time_stride = m_chan_g * norm_chan_stride;

    const int wt_chan_stride  = n_pol;
    const int wt_bl_stride    = n_vis_chan * wt_chan_stride;
    const int wt_time_stride  = n_baseline * wt_bl_stride;

    const int uvw_bl_stride   = 3;
    const int uvw_time_stride = n_baseline * uvw_bl_stride;

    for (int i_time = 0; i_time < n_time; ++i_time) {
        const int a_time   = time_map[i_time];
        const int wt_t_off = i_time * wt_time_stride;
        const int uvw_t_off = i_time * uvw_time_stride;

        for (int i_baseline = 0; i_baseline < n_baseline; ++i_baseline) {
            const int wt_b_off  = wt_t_off  + i_baseline * wt_bl_stride;
            const int uvw_b_off = uvw_t_off + i_baseline * uvw_bl_stride;

            for (int i_chan = 0; i_chan < n_vis_chan; ++i_chan) {
                const int a_chan = frequency_map[i_chan];

                const double u = uvw[uvw_b_off + 0] * uv_scale_u[i_chan];
                const double v = uvw[uvw_b_off + 1] * uv_scale_v[i_chan];

                if (std::isnan(u) || std::isnan(v)) continue;

                const double u_pos = u + uv_center_u;
                const double v_pos = v + uv_center_v;

                const int u_center_indx = static_cast<int>(u_pos + 0.5);
                const int v_center_indx = static_cast<int>(v_pos + 0.5);

                if (u_center_indx + support_center >= m_u) continue;
                if (v_center_indx + support_center >= m_v) continue;
                if (u_center_indx - support_center <  0)   continue;
                if (v_center_indx - support_center <  0)   continue;

                const double u_offset = u_center_indx - u_pos;
                const int    u_center_offset_indx =
                    static_cast<int>(std::floor(u_offset * oversampling + 0.5));

                const double v_offset = v_center_indx - v_pos;
                const int    v_center_offset_indx =
                    static_cast<int>(std::floor(v_offset * oversampling + 0.5));

                const int wt_c_off = wt_b_off + i_chan * wt_chan_stride;

                for (int i_pol = 0; i_pol < n_pol; ++i_pol) {
                    const double weight_data = weight[wt_c_off + i_pol];

                    if (std::isnan(weight_data) || weight_data == 0.0) continue;

                    const int a_pol = pol_map[i_pol];

                    const int grid_base = a_time * grid_time_stride
                                        + a_chan  * grid_chan_stride
                                        + a_pol   * grid_pol_stride;
                    const int norm_idx  = a_time * norm_time_stride
                                        + a_chan  * norm_chan_stride
                                        + a_pol;

                    double norm = 0.0;

                    for (int i_v = start_support; i_v < end_support; ++i_v) {
                        const int v_indx = v_center_indx + i_v;
                        const int v_offset_indx =
                            std::abs(oversampling * i_v + v_center_offset_indx);
                        const double conv_v = cgk_1D[v_offset_indx];

                        for (int i_u = start_support; i_u < end_support; ++i_u) {
                            const int u_indx = u_center_indx + i_u;
                            const int u_offset_indx =
                                std::abs(oversampling * i_u + u_center_offset_indx);
                            const double conv = cgk_1D[u_offset_indx] * conv_v;

                            grid[grid_base + u_indx * m_v + v_indx] +=
                                std::complex<double>(conv * weight_data, 0.0);
                            norm += conv;
                        }
                    }

                    normalization[norm_idx] += weight_data * norm;
                }
            }
        }
    }
}

} // namespace prolate_spheroidal
